### ISSUE TITLE
Timeout on enquiries GET endpoint when sorting or filtering by `modified`

### ISSUE THREAD

**Issue by tomgidden** [reapit/foundations#8910]
Trying to retrieve `GET /enquiries` with `sortBy=modified` **OR** `modifiedFrom=...` / `modifiedTo=...` times out with a `504 Gateway Timeout` error.

* Customer: `RBJ`;  issue doesn't affect `SBOX`.
* Same code works on `contacts` and `applicants`; but `enquiries` has _many_ more records.
* Same code works when run without `sortBy=` or with `sortBy=created`.
* Unaffected by `pageSize=`

So, this works:
```
curl -H 'accept: application/json' -H 'api-version: 2020-01-31' -H 'reapit-customer: RBJ' -H "Authorization: Bearer $TOKEN" \
 'https://platform.reapit.cloud/enquiries/'
```

but this times out:
```
curl -H 'accept: application/json' -H 'api-version: 2020-01-31' -H 'reapit-customer: RBJ' -H "Authorization: Bearer $TOKEN" \
 'https://platform.reapit.cloud/enquiries/?sortBy=modified'
```

For what it's worth, it feels like there's a missing index on the `modified` column.

**Specification**
Experimental change
- Project: Reapit.Services.Enquiries
- Work with SI to check index on appimp.aAmend for this customer only for the time being
- Switch query mechanism to use a union instead of coalesce. Currently when using the modifiedFrom parameter the resultant query, to account for the amend date not being set on new records is in the format:

```
SELECT * FROM tbl WHERE COALESCE(register, amend) >='2023-04-01'
```
This causes MySQL to ignore the indexes however a workaround is to use a union as follows, which will allow MySQL to use the indexes resulting in much better performance. This is done in EF by using `.Concat` NOT `.Union`

```
SELECT * FROM tbl WHERE register >='2023-04-01'
UNION
SELECT * FROM tbl WHERE amend >='2023-04-01'
```

**Comment by plittlewood-rpt** [comment]
Hi @tomgidden this is a known issue and something that we are actively working on. There are some changes being made to the core CRM product that will allow us to alter how we run the sort at the database layer which will result in much faster performance. As it's a core product change it is taking a while to get out but we are very nearly there now. I've assigned this to our performance milestone and will leave the item open so I can feed back to you when the updates are out in production. Are you seeing this timeout persistently or is it intermittent?

**Comment by tomgidden** [comment]
Hi @plittlewood-rpt... unfortunately it's 100% consistent for `sortBy=modified` or just a plain `modifiedFrom=` query, from what I can tell.

However, it does seem to work intermittently when I do `sortBy=created&modifiedFrom=...` providing the `modifiedFrom=....` date window is small enough. I expect this is still doing a full table-scan for the filter (slow) but then NOT having to do a sort that would end up causing a timeout.

That's not ideal, but it does give me a possible workaround where as long as the code performs frequent updates and will retry sensibly with backing off when it detects Gateway Timeouts, it may be possible to maintain an up-to-date index of the required data. It'll work for a proof-of-concept for a single client, but I doubt it can be relied upon for production.  Hopefully your changes will come in time for that, though.

Thanks for the update!

**Comment by plittlewood-rpt** [comment]
Hi @tomgidden  thanks - I've chased our desktop product owner for an update on where we are at with getting the changes out that will allow us to resolve this fully. In the meantime, it might help to understand your precise use case so I can evaluate further. Would, for example, it be of any benefit if we added additional query string parameters to the API to allow you to strip out enquiries you're not interested in? If so that would probably allow us to use other indexes on the table which might improve things for you.

**Comment by tomgidden** [comment]
Hi @plittlewood-rpt... thanks. This is for cross-referencing submitted forms in our DB as well as analytics data with Reapit data.  As we're doing some fuzzy/inexact matches, it's not really practical to query Reapit directly -- they'd be complicated searches, and I expect a prohibitive number of API hits.

So instead, I'm keeping an up-to-date copy of all the objects, and redacting data once I know it's no longer relevant. So for this, I need query everything modified since the last update; ie. `modifiedFrom=$MOST_RECENT_MODIFIED_DATE`.  Sorting by `+modified` would also be useful so partial errors have less effect, but those seem to have the same issue, so I just sort them all on this side once I know we've got a complete successful batch.

I've found that as long as the script runs frequently enough so the window is small enough -- about an hour -- and the API hit correctly retries on a Gateway Timeout error, it'll get it in 3 or 4 goes.  As long as the script continues to run regularly, it should be okay.

Some extra things, like filtering out different enquiry types as you suggest, would be useful, but I think the table is big enough that it'll still probably timeout on occasion without a working `modified` index.
